File: day-09/python/part2.py
#!/usr/bin/env python3

import sys

data = [int(x.strip()) for x in open(sys.argv[1], "r").readlines()]

step = int(sys.argv[2])

def find_invalid(step, data):
    for i in range(step, len(data)):
        found = False
        for j in range(i-step, i):
            for k in range(i-step, i):
                if data[i] == data[k] + data[j]:
                    found = True
        if not found:
            return data[i]

invalid_num = find_invalid(step, data)

s = []

for i in range(len(data)):
    for j in range(i+1, len(data)):
        if sum(data[i:j+1]) == invalid_num:
            for x in data[i:j+1]:
                s.append(x)
            break
        if sum(data[i:j+1]) > invalid_num:
            break

s.sort()
print(s[0]+s[-1])

File: day-09/nim/part2.nim
from os import paramStr
from algorithm import sort
import sugar, strutils, sequtils

let 
  data = collect(newSeq):
    for x in lines(paramStr(1)):
      parseInt(x.strip())
  step = parseInt(paramStr(2))

proc find_invalid(step: int, data: seq[int]): int = 
  for i in step ..< data.len:
    var found = false
    for j in (i - step) ..< i:
      for k in (i - step) ..< i:
        if k!=j and data[i] == data[k] + data[j]:
          found = true
    if not found:
      return data[i]

let invalid_num = find_invalid(step, data)

var s = newSeq[int](0)

for i in 0 ..< data.len:
  for j in i+1 ..< data.len:
    if data[i ..< j+1].foldl(a+b) == invalid_num:
      for x in data[i ..< j+1]:
        s.add(x)
    if data[i ..< j+1].foldl(a+b) > invalid_num:
      break

s.sort
echo s[0]+s[^1]
